update: improve output in debug/verbose mode

UniqMartin opened the pull request:
Running `brew update --verbose`, particularly with many taps, currently produces a gibberish of intermixed output due to parallel fetching (See [update-v-parallel.txt](https://gist.github.com/UniqMartin/e63898f3122f1e817470#file-update-v-parallel-txt) for an example.) This PR addresses this in its 2nd commit by switching to serial fetching in debug/verbose mode, assuming that a clear output is more important than performance. (See [update-v-serialized.txt](https://gist.github.com/UniqMartin/e63898f3122f1e817470#file-update-v-serialized-txt) for the improved output.)

The 3rd commit improves on that (and in a way addresses #49893) by adding a header before every fetch and update (merge or rebase) operation as well as before handing over to `brew update-report`. (Of course only in verbose mode. See [update-v-with-headers.diff](https://gist.github.com/UniqMartin/e63898f3122f1e817470#file-update-v-with-headers-diff) for a diff that highlights the added headers.)

The 1st commit provides previously not available shell implementations of `ohai`, `oh1`, etc. the above changes depend on. (I thoroughly tested things locally, but the changes are actually fairly non-intrusive.)

mikemcquaid commented:
I'd rather silence the output than make it non-parallel. It feels really weird for `--verbose` to also imply parallelisation be disabled.

mikemcquaid commented on Library/brew.sh:
```diff
@@ -1,16 +1,46 @@
+
+odie() {
+  # Use underline and red color if `stderr` is a TTY.
+  styled-output 2 "Error: %s" "\033[4;31mError\033[0m: %s" "$@" >&2
```
Feels like we're starting to warrant a `utils.sh` or similar for all of these.

UniqMartin replied:
Yes, that makes a lot of sense! I just wanted to get them implemented and reviewed first before worrying about placement. I'm also not entirely sure if they should live in `Library/Homebrew` or somewhere else? It feels both weird and natural to put them next to their Ruby counterparts.

UniqMartin commented:
> I'd rather silence the output than make it non-parallel. It feels really weird for `--verbose` to also imply parallelisation be disabled.

Personally, I don't care that much about the verbose output (I almost never run `brew update` like that), but I figured since we offer a verbose mode, it should produce an output that can be made sense of. Disabling parallelization was the easiest way to achieve that, and given that this is opt-in behavior, possibly the most pragmatic way to solve that. (One could do more elaborate things like capture the outputs of those `git` processes and then print them in a coordinated fashion, something I'd rather not implement in Bash.)

As for taking away the verbosity to be able to keep parallelism, I suspect some people will be mad that we're not giving them the detail they were used to have. (And that despite the garbled output in the `fetch` phase and the difficult to interpret output in the `merge` phase. What repository does the output belong to?) I guess in a few rare instances it's actually helpful to see the output to debug a (connectivity) issue.

After all, the default is still a parallel (and silent) fetch. What is your primary concern here? That many people have set `HOMEBREW_VERBOSE` in their environment and thus will lose parallel fetch? (Note that there's a warning being printed informing them about the performance implications.)

chdiza commented:
FWIW, I don't see (and have never seen) the reported intermingling of output when I run `brew update` not with the CLI flag, but with `HOMEBREW_VERBOSE` set.

UniqMartin commented:
> FWIW, I don't see (and have never seen) the reported intermingling of output when I run `brew update` not with the CLI flag, but with `HOMEBREW_VERBOSE` set.

Whether you're using `--verbose` or `HOMEBREW_VERBOSE` isn't relevant. They should give the same result. The intermingling can only happen if there are multiple Git repositories being fetched in parallel. You mentioned elsewhere that you don't have any taps you've tapped from a remote location, so your observation is not surprising. Depending on network speed, the number of repositories being fetched, and whether there are any changes to be fetched (among other factors), output of concurrently running `git fetch` instances will more or less likely happen at the same time. (Due to non-determinism involved here, the exact output can be different each time.)

The linked output I provided in my initial comment, that shows the intermingling, is from an installation where the taps haven't been updated in three weeks and there are about two dozen of them.

mikemcquaid commented:
> After all, the default is still a parallel (and silent) fetch. What is your primary concern here? That many people have set HOMEBREW_VERBOSE in their environment and thus will lose parallel fetch? (Note that there's a warning being printed informing them about the performance implications.)

My main concern is that `--verbose` implying `--no-parallel` or similar is unintuitive. I'd just focus here on fixing @chdiza's specific case (but, honestly, for extremely niche side-cases like this I think @chdiza is probably able to fix the issue themselves).

xu-cheng commented:
I would suggest just add `oh1 "Homebrew update report" if ARGV.verbose?` in the beginning of `update-report.rb`.

chdiza commented on Library/Homebrew/cmd/update.sh:
```diff
@@ -364,6 +389,10 @@ EOS
+  if [[ -n "$HOMEBREW_VERBOSE" ]]
+  then
+    oh1 "Summarizing Homebrew update"
+  fi
```
I would suggest that this message end with a colon. Also, I hope that this last message will be shown whether or not the user has any taps at all, and whether or not the taps are git repos, and that there not be a "heading-less" emission of "Already up-to-date" in any of those same circumstances.

I still think the best solution is to drop this header, but make sure the reported message isn't identical to `Already up-to-date`.  Maybe `Homebrew installation is already up to date` or something of that nature.

UniqMartin commented:
I hope I was able to address all feedback and still improve the output in verbose mode, which was my actual goal. This is now also reflected in the updated issue title, as the previous put too much focus on disabling parallelization; this was just a side effect. In summary (see individual commits for details):

- Introduce shell versions of `oha1`, `oh1`, `opoo`, and `onoe` as basis for all the other changes. (Now relocated to `Library/Homebrew/utils.sh`.)

- Allow to disable parallel fetch via `--no-parallel` as a basis for the next change. (Undocumented for now, but I'm happy to add documentation if the flag is accepted.)

- If the user runs `brew update --verbose`, tell them that output might be less useful during the fetch phase and provide a remedy with caveats (the `--no-parallel` option), but proceed in parallel.

- *(This is the primary goal of this PR and the part I care most about.)* In verbose mode, add headings that indicate the current phase (fetch, update, report) and, if applicable, the repository the output belongs to (see example below), which provides much needed context to make sense of the output.

----

Sample output of an installation with only one tap, in this case a no-op update. Changes in the output are highlighted as a pseudo-diff. (The headings are obviously much more useful if there is something to fetch and merge, at which point they will no longer dominate the output, but provide helpful context.)

```diff
 $ brew update --verbose
+Warning: Enabling debug and/or verbose mode can lead to overlapping output while
+updates are fetched in parallel. If this output is important to you, retry with:
+  brew update --no-parallel --verbose
+
+Note that disabling parallel fetch tends to significantly slow down the update.
+==> Fetching updates for all Git repositories
+==> Updating all Git repositories
+==> Updating /opt/brewery/petsc
 Switched to branch 'master'
 Your branch is up-to-date with 'origin/master'.
 Already up-to-date.
+==> Updating /opt/brewery/petsc/Library/Taps/homebrew/homebrew-science
 Already up-to-date.
-Already up-to-date.
+==> Homebrew update report
+Homebrew is already up-to-date.
```

The same example with `--no-parallel` added.

```diff
 $ brew update --no-parallel --verbose
+==> Fetching updates for all Git repositories
+==> Fetching updates for /opt/brewery/petsc
+==> Fetching updates for /opt/brewery/petsc/Library/Taps/homebrew/homebrew-science
+==> Updating all Git repositories
+==> Updating /opt/brewery/petsc
 Switched to branch 'master'
 Your branch is up-to-date with 'origin/master'.
 Already up-to-date.
+==> Updating /opt/brewery/petsc/Library/Taps/homebrew/homebrew-science
 Already up-to-date.
-Already up-to-date.
+==> Homebrew update report
+Homebrew is already up-to-date.
```

mikemcquaid commented on Library/Homebrew/cmd/update.sh:
```diff
@@ -312,6 +318,20 @@ EOS
+Note that disabling parallel fetch tends to significantly slow down the update.
+EOS
+    fi
+  fi
```
Personally I think it's worth omitting this until it's requested/complained about.

mikemcquaid replied:
As it creates another option (and possible edge cases) that we'll have to support.

UniqMartin replied:
Agreed.

mikemcquaid commented on Library/Homebrew/cmd/update-report.rb:
```diff
@@ -45,7 +47,7 @@ def update_report
 
     if hub.reporters.empty?
-      puts "Already up-to-date."
+      puts "Homebrew is already up-to-date."
```
Would probably rather leave this as-is.

UniqMartin replied:
Makes sense, particularly given that this affects a lot more users than the presumably few that default to verbose output. And the previous ambiguity as to what this refers to should be gone, now that there's a title printed above (in verbose mode only). Thanks for staying critical!

mikemcquaid commented on Library/Homebrew/cmd/update.sh:
```diff
@@ -321,6 +341,11 @@ EOS
 
+  if [[ -n "$HOMEBREW_VERBOSE" ]]
+  then
+    oh1 "Fetching updates for all Git repositories"
```
Feels like this is unnecessary given it'll be printed for each of the `$DIR` below?

UniqMartin replied:
The output for each `$DIR` below only happens in non-parallel mode (notice the additional condition there), as it would otherwise just contribute to the intermingling of output. Thus this is the only line always being printed during the fetch phase. And since I'm about to rip out the `--no-parallel` stuff for now, the repository-specific output will have to go and this line probably should stay.

mikemcquaid replied:
:ok: :cool:

UniqMartin commented:
Alright, `--no-parallel` and everything related to it is gone (until there's some evidence it is actually needed and useful) and other minor comments should have been addressed, too. I also slightly tidied up the code and added a few `shellcheck` comments to silence irrelevant warnings.

What remains are additional helpers to print various types of formatted messages to the terminal and (in verbose mode) a few extra lines of output in `update` and `update-report` that provide context and help to make sense of the output.

Here's what a verbose no-op update will look like if the PR is merged as-is (pseudo-diff again):

```diff
 $ brew update --verbose
+==> Fetching updates for all Git repositories
+==> Updating all Git repositories
+==> Updating /opt/brewery/petsc
 Switched to branch 'master'
 Your branch is up-to-date with 'origin/master'.
 Already up-to-date.
+==> Updating /opt/brewery/petsc/Library/Taps/homebrew/homebrew-science
 Already up-to-date.
+==> Homebrew update report
 Already up-to-date.
```

mikemcquaid commented:
> +==> Fetching updates for all Git repositories
> +==> Updating all Git repositories

I'm still not sure I understand the need for these two headers if there's the update headers below?

mikemcquaid commented on Library/Homebrew/utils.sh:
```diff
@@ -0,0 +1,47 @@
+
+  if [[ -t "$output_fd" ]] # If FD is a TTY, style message with color/bold/...
+  then
+    # shellcheck disable=SC2059
```
Can you add a comment to explain why?

UniqMartin replied:
[The warning being silenced here](https://github.com/koalaman/shellcheck/wiki/SC2059) is warranted when using user-provided or generated content as the format string for `printf`, but in our case the format string is fixed (coming from `ohai`, `oh1`, etc.), so the warning doesn't apply.

mikemcquaid replied:
Inline comment would be great. Could you just use e.g. `echo` to avoid the warning?

UniqMartin replied:
Yes, I can put that explanation into the code. We can't use `echo` here because the message goes in the middle of a format string that (for the typical case of a TTY) contains escape sequences for styling before and/or after the `%s` placeholder.

mikemcquaid commented on Library/Homebrew/utils.sh:
```diff
@@ -0,0 +1,47 @@
+    message="$*"
+  fi
+
+  if [[ -t "$output_fd" ]] # If FD is a TTY, style message with color/bold/...
```
Stick comments on the line above.

UniqMartin replied:
:ok:

mikemcquaid commented on Library/Homebrew/utils.sh:
```diff
@@ -0,0 +1,47 @@
+
+ohai() {
+  # Use bold and blue color if `stdout` is a TTY.
+  puts_with_style 1 "==> %s" "\033[1;34m==>\033[1;39m %s\033[0m" "$@"
```
Not obvious to me what the `1` refers to.

UniqMartin replied:
It should become obvious when looking at the first line of `puts_with_style` where the first argument gets assigned to `output_fd`. Would it be sufficiently obvious if I modified the comment above this line to `Use bold and blue color if FD 1 (stdout) is a TTY.`?

mikemcquaid replied:
Probably ideal to have a new variable for `STDOUT_FD` or something

UniqMartin replied:
I'm not sure what you mean by that. Can you elaborate?

My impression was that stuff like `2>&1` is used in almost every shell script and is thus pretty well understood, hence: that FD 1 stands for `stdout` and FD 2 for `stderr` is common knowledge. Thus hinting at the fact that the first argument is an FD could/should be sufficient?!

mikemcquaid replied:
It's generally obvious but it's not obvious that this parameter is a FD.

UniqMartin replied:
So what do I do? The best I was able to do in `puts_with_style` was to immediately assign the first argument to a local variable named `output_fd` and to put this function just above the others and well within sight …

One could imagine an interface like `puts_with_style --fd=1 --plain-format="==> %s" --tty-format="\033[1;34m==>\033[1;39m %s\033[0m" -- <message>` but that feels a bit overengineered for a simple function that is only used internally to reduce code duplication among `odie` and friends.

mikemcquaid replied:
```bash
STDOUT="1"
puts_with_style "$STDOUT"
...
STDERR="2"
puts_with_style "$STDERR"
```

mikemcquaid commented on Library/brew.sh:
```diff
@@ -1,20 +1,7 @@
-  fi
-  exit 1
-}
+# shellcheck source=/dev/null
```
Can you explain this comment?

UniqMartin replied:
`shellcheck` complains that it cannot check the sourced file because it cannot find it (because the path is using some variable and its value is only known at run time). In theory, this could be:

```
# shellcheck source=Library/Homebrew/utils.rb
```

But then `shellcheck` will print an even more confusing warning that it can't find the file, unless one always makes sure to run `shellcheck` with `$HOMEBREW_REPOSITORY` as the current working directory.

mikemcquaid replied:
Ok, cool. Another inline comment would be handy.

UniqMartin replied:
That's a tough one to condense into an inline comment, but I'll try …

UniqMartin commented:
> > +==> Fetching updates for all Git repositories
> > +==> Updating all Git repositories
>
> I'm still not sure I understand the need for these two headers if there's the update headers below?

It's verbose mode, so an additional header or two shouldn't hurt and the first one has the benefit of being printed before hitting the network, so there's some output and hence a feeling of progress.

Furthermore, the idea is that the update consists of three main phases (fetch, update, report) that are styled as `oh1`. Those can have sub-phases that are styled as `ohai`. The per-repository headers fall into the second category (and are necessary for the output to have enough context to make sense).

Maybe this tree representation of the (existing and new) headers and their relationship helps:

```
$ brew update --verbose
|-- Fetching updates for all Git repositories
|-+ Updating all Git repositories
|  |-- Updating /opt/brewery/petsc
|  \-- Updating /opt/brewery/petsc/Library/Taps/homebrew/homebrew-science
\-+ Homebrew update report
  |-- Added Formulae
  |-- Updated Formulae
  \-- Deleted Formulae
```

mikemcquaid commented:
> |-- Fetching updates for all Git repositories

This is a section heading with no contents which is why I think it should go.

> |-+ Updating all Git repositories

This is a section heading that will always have entries which is why I think it should go.

UniqMartin commented:
Somehow I'm getting the feeling that my example output was poorly chosen or that I'm just too stupid to communicate clearly (or worse yet, that my thoughts just don't make any sense). :crying_cat_face: 

> > |-- Fetching updates for all Git repositories
>
> This is a section heading with no contents which is why I think it should go.

Of course this heading has contents following it, namely the output of all the `git fetch` instances that will dump their output between this and the other heading, as long as there is something to fetch.

> > |-+ Updating all Git repositories
>
> This is a section heading that will always have entries which is why I think it should go.

But if I remove that top-level heading the structure I tried to explain doesn't make sense anymore and its children (the per-repository headings and their individual `git merge` outputs) are no longer recognizable as sub-elements of the 2nd phase (update, a.k.a merge/rebase).

mikemcquaid commented:
> as long as there is something to fetch.

This is the bit that had me confused. In that case perhaps `Checking for updates to Homebrew and taps`

mikemcquaid commented:
> But if I remove that top-level heading the structure I tried to explain doesn't make sense anymore and its children (the per-repository headings and their individual git merge outputs) are no longer recognizable as sub-elements of the 2nd phase (update, a.k.a. merge/rebase).

Ok. Perhaps instead print `Checking for updates to Homebrew` or `Checking for updates to $TAP_NAME`

mikemcquaid commented:
I'm sorry if I'm being pedantic here I just think it's worth thinking about this stuff and I'm still not sure I think the additional complexity is worth it rather for @chdiza's one-line complaint.

apjanke commented:
This is probably obsolete due to the recent changes to `brew update`, or at least needs to be re-submitted against the head of `homebrew/brew`, yes?

UniqMartin commented:
> This is probably obsolete due to the recent changes to `brew update`, or at least needs to be re-submitted against the head of `homebrew/brew`, yes?

It's not obsolete. I simply need to address some comments before I resubmit this to `Homebrew/brew`. It got prioritized down a bit, but I'll get back to this—hopefully soon.

apjanke commented:
Cool. Closing this out so we can clear out legacy-homebrew then, since it's not mergeable as-is.
